internal/tools/terraformlinter/terraform_linter.go

```diff
@@ -0,0 +1,335 @@
+	attrCount                  = "count"
+	attrProvider               = "provider"
+	attrSource                 = "source"
+	attrProviderProject        = "project"
+	attrProviderProjectID      = "project_id"
+	attrProviderFolder         = "folder"
+	attrProviderFolderID       = "foler_id"
+	attrProviderOrganization   = "organization"
+	attrProviderOrganizationID = "organization_id"
+	attrProviderOrgID          = "org_id"
+	attrDependsOn              = "depends_on"
+	attrLifecycle              = "lifecycle"
+)
+
+// mapping of attributes to their expected position.
+var positionMap = map[string]tokenPosition{
+	attrForEach:                Leading,
+	attrCount:                  Leading,
+	attrProvider:               Leading,
+	attrSource:                 Leading,
+	attrProviderProject:        Provider,
+	attrProviderProjectID:      Provider,
+	attrProviderFolder:         Provider,
+	attrProviderFolderID:       Provider,
+	attrProviderOrganization:   Provider,
+	attrProviderOrganizationID: Provider,
+	attrProviderOrgID:          Provider,
+	attrDependsOn:              Trailing,
+	attrLifecycle:              Trailing,
+}
+
+const (
+	errorLeadingMetaBlockAttribute  = "The attribute %q must be in the meta block at the top of the definition."
+	errorMetaBlockNewline           = "The meta block must have an additional new line separating it from the next section."
+	errorProviderAttributes         = "The attribute %q must me below any meta attributes (for_each, count, etc.) but above all other attributes."
+	errorProviderNewline            = "The provider specific attributes must have an additional new line separating it from the next section."
+	errorTrailingMetaBlockAttribute = `The attribute %q must be at the bottom of the resource definition and in the order "depends_on" then "lifecycle."`
+)
+
+// ViolationInstance is an object that contains a reference to a location
+// in a file where a lint violation was detected.
+type ViolationInstance struct {
+	ViolationType string
+	Path          string
+	Line          int
+}
+
+// RunLinter executes the specified linter for a set of files.
+func RunLinter(ctx context.Context, paths []string) error {
+	var violations []*ViolationInstance
+	// Process each provided path looking for violations
+	for _, path := range paths {
+		instances, err := lint(path)
+		if err != nil {
+			return fmt.Errorf("error linting files: %w", err)
+		}
+		violations = append(violations, instances...)
+	}
+	for _, instance := range violations {
+		fmt.Printf("%q detected at [%s:%d]\n", instance.ViolationType, instance.Path, instance.Line)
+	}
+	if len(violations) != 0 {
+		return fmt.Errorf("found %d violation(s)", len(violations))
+	}
+
+	return nil
+}
+
+// lint reads a path and determines if it is a file or a directory.
+// When it finds a file it reads it and checks it for violations.
+// When it finds a directory it calls itself recursively.
+func lint(path string) ([]*ViolationInstance, error) {
```

Comment:
Not required change, just an FYI. If you accept an `fs.FS`, it makes testing easier because you can create an in-memory FS pretty easily.

Reply:
I'll keep that in mind for the future. Right now its a string on the command line and filepath.WalkDir wants a string so converting seems wasteful.